panda/src/navigation/navMesh.cxx

```diff
@@ -0,0 +1,388 @@
+    dg.add_uint32(_params.detailMeshes[i]);
+  }
+
+  //detail mesh vertices
+  for(int i=0 ; i < (_params.detailVertsCount) * 3 ;i++) {
+    dg.add_float32(_params.detailVerts[i]);
+  }
+
+  //detail mesh vertices
+  for(int i=0 ; i < (_params.detailTriCount) * 4 ;i++) {
+    dg.add_uint8(_params.detailTris[i]);
+  }
+
+  //RC_MESH_NULL_IDX
+  dg.add_int32(RC_MESH_NULL_IDX);
+
+}
+
+/**
+ * This function is called by the BamReader's factory when a new object of
+ * type NavMesh is encountered in the Bam file.  It should create the
+ * NavMesh and extract its information from the file.
+ */
+TypedWritable *NavMesh::
+make_from_bam(const FactoryParams &params) {
+  NavMesh *param = new NavMesh;
+  DatagramIterator scan;
+  BamReader *manager;
+
+  parse_params(params, scan, manager);
+  param->fillin(scan, manager);
+
+  return param;
+}
+
+/**
+ * This internal function is called by make_from_bam to read in all of the
+ * relevant data from the BamFile for the new NavMesh.
+ */
+void NavMesh::
+fillin(DatagramIterator &scan, BamReader *manager) {
+  //NavMesh::fillin(scan, manager);
+  
+  int vert_count = scan.get_int32();
+  int poly_count = scan.get_int32();
+  int nvp = scan.get_int32();
+  int detail_vert_count = scan.get_int32();
+  int detail_tri_count = scan.get_int32();
+  float walkable_height = scan.get_float32();
+  float walkable_radius = scan.get_float32();
+  float walkable_climb = scan.get_float32();
+  float cs = scan.get_float32();
+  float ch = scan.get_float32();
+  bool build_bv_tree = scan.get_bool();
+
+  float b_min[3];
+  for(int i=0;i<3;i++) {
+    b_min[i] = scan.get_float32();
+  }
+  float b_max[3];
+  for(int i=0;i<3;i++) {
+    b_max[i] = scan.get_float32();
+  }
+
+  //POLYGON MESH ATTRIBUTES
+
+  unsigned short *verts = new unsigned short[3 * vert_count];
+  for(int i=0 ; i < 3 * vert_count ; i++) {
+    verts[i] = scan.get_uint16();
+  }
+
+  unsigned short *polys = new unsigned short[poly_count * 2 * nvp];
+  for(int i=0 ; i < poly_count * 2 * nvp ; i++) {
+    polys[i] = scan.get_uint16();
+  }
+
+  unsigned short *poly_flags = new unsigned short[poly_count];
+  for(int i=0 ; i < poly_count; i++) {
+    poly_flags[i] = scan.get_uint16();
+  }
+
+  unsigned char *poly_areas = new unsigned char[poly_count];
+  for(int i=0 ; i < poly_count; i++) {
+    poly_areas[i] = scan.get_uint8();
+  }
+
+  //POLYGON MESH DETAIL ATTRIBUTES
+
+  unsigned int *detail_meshes = new unsigned int[poly_count * 4];
+  for(int i=0 ; i < poly_count * 4 ;i++) {
+    detail_meshes[i] = scan.get_uint32();
+  }
+
+  float *detail_verts = new float[detail_vert_count * 3];
+  for(int i=0 ; i < detail_vert_count * 3 ;i++) {
+    detail_verts[i] = scan.get_float32();
+  }
+
+  unsigned char *detail_tris = new unsigned char[detail_tri_count * 4];
+  for(int i=0 ; i < detail_tri_count * 4 ;i++) {
+    detail_tris[i] = scan.get_uint8();
+  }
+
+  //RC_MESH_NULL_IDX
+  RC_MESH_NULL_IDX = scan.get_int32();
```

Comment:
This is a constant, we don't need to store it anywhere.

Reply:
@rdb the vertices with this value belong to the edges which are solid borders. So if someone does not use navmeshgen to build the navmesh, then he/she might have a different value for such vertices.

Reply:
Renamed it